src/modules/package.rs

```diff
@@ -88,25 +90,75 @@ fn get_jsr_package_version(context: &Context, config: &PackageConfig) -> Option<
     format_version(raw_version, config.version_format)
 }
 
-fn get_poetry_version(pyproject: &toml::Table) -> Option<&str> {
+fn get_poetry_version(pyproject: &toml::Table) -> Option<String> {
     pyproject
         .get("tool")?
         .get("poetry")?
         .get("version")?
         .as_str()
+        .map(|s| s.to_owned())
 }
 
-fn get_pep621_version(pyproject: &toml::Table) -> Option<&str> {
-    pyproject.get("project")?.get("version")?.as_str()
+fn parse_file_version_for_hatchling(context: &Context, path: &str) -> Option<String> {
+    let file_contents = read_file(context.current_dir.join(path)).ok()?;
+    // https://hatch.pypa.io/latest/version/
+    let re = Regex::new(r#"(__version__|VERSION)\s*=\s*["']([^"']+)["']"#).ok()?;
+    Some(
+        re.captures(&file_contents)
+            .and_then(|cap| cap.get(2))?
+            .as_str()
+            .to_owned(),
+    )
+}
+
+fn parse_hatchling_dynamic_version(context: &Context, pyproject: &toml::Table) -> Option<String> {
+    let version_path = pyproject
+        .get("tool")?
+        .get("hatch")?
+        .get("version")?
+        .get("path")?
+        .as_str()?;
+
+    parse_file_version_for_hatchling(context, version_path)
+        .and_then(|s| Version::new(s.as_str()))
+        .map(|v| v.to_string())
```

Comment:
Apart from that LGTM if tests pass once `master` is fixed.

```suggestion
        .filter(|s| Version::new(s.as_str()).is_ok())
```

Reply:
The `filter` maps the value to `None` if the it can't parse the `Version`, while not changing the original value if it can. Thus the original verified string can be returned because it is still there.